### DISCUSSION TITLE
How to make a custom ASR architecture in espnet2?

### DISCUSSION THREAD

**Discussion by geoffvdr** [espnet/espnet#4954] [Q&A]
Hi everyone,

I am relatively new to espnet. I am experimenting with the encoder-decoder architecture in espnet2.asr.espnet_model.
I would like to add a simple 2 layer feed forward neural network between the encoder and the decoder;
is there a simple way to do that or do I have to rewrite/adapt the espnet2.asr.espnet_model ?
I found that with espnet2.asr.postencoder.hugging_face_transformers_postencoder I can add a huggingface model as postencoder, however I would like to start with something simple where I can define the architecture myself.

In general, what is the good approach when you want to implement a new architecture (say combine multiple encoders/decoders, add components like a feed-forward neural network)?

Thank you for your help.

Geoff

**Comment by sw005320** [comment]
If you do something between the encoder and decoder networks, you can check [this part](https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L221-L316) and modify it 

`encoder_out` is extracted from an encoder in https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L221, and then used at a decoder in https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L316
So, if you want to add a simple 2-layer FFN, you can add it around.
Same for combining multiple encoders/decoders. This can be implemented in espnet_model.py

**Reply by geoffvdr** [reply]
Thank you @sw005320, that is indeed what I thought. And so where do you specify this new espnet_model.py in the asr template then? Do you need to make a new bin to replace this line https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh#L1248 ?

**Reply by sw005320** [reply]
If you add an option for this new implementation and changing the corresponding config file name, the experiments will be done smoothly without changing the other parts.
Then, you can redo the experiments with this new config from stage 10 or 11.

**Reply by geoffvdr** [reply]
Thanks a lot!
